File: C++/20.观察者模式/Subject.cpp
/**
 * @file Subject.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-21
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "Subject.h"

Subject::Subject() {}

Subject::~Subject() { observer_list_.clear(); }

void Subject::SetState(const int state) {
    state_ = state;
    NotifyAllObservers();
}

int Subject::GetState() { return state_; }

void Subject::Attach(Observer* observer) { observer_list_.push_back(observer); }

void Subject::NotifyAllObservers() {
    for (auto iter : observer_list_) {
        iter->Update();
    }
}

File: Objective-C/20.ObserverPattern/Subject.m
//
//  Subject.m
//  20.ObserverPattern
//
//  Created by Layne on 2020/11/13.
//

#import "Subject.h"
#import "Observer.h"

@implementation Subject {
    int _state;
    NSMutableArray* _observers;
}

- (int)getState {
    return _state;
}
- (void)setState:(int)state {
    _state = state;
    [self notifyAllObservers];
}

- (void)attach:(Observer*)observer {
    if (_observers == nil) {
        _observers = [[NSMutableArray alloc] init];
    }
    [_observers addObject:observer];
}

- (void)notifyAllObservers {
    for (Observer* obsever in _observers) {
        [obsever update];
    }
}

@end
